feat: UI Dialog and DialogTrigger Components (#953)
deephaven/deephaven-plugins: plugins/ui/docs/components/dialog.md

# Dialog

> Dialogs are windows containing contextual information, tasks, or workflows that appear over the user interface. Depending on the kind of Dialog, further interactions may be blocked until the Dialog is acknowledged.

Reviewer: I don't think Dialog needs to be capitalized.

Author: Updated

                    gap="size-100",
                )
            ),
            ui.content(
                ui.form(
                    ui.text_field(label="First Name", auto_focus=True),
                    ui.text_field(label="Last Name"),
                    ui.text_field(label="Street Address"),
                    ui.text_field(label="City"),
                    ui.checkbox(
                        "I want to receive udpates for exclusive offers in my area."
                    ),
                )
            ),
            ui.button_group(
                ui.button("Cancel", variant="secondary", on_press=set_closed),
                ui.button("Register", variant="accent", on_press=set_closed),
            ),
        ),
        is_open=is_open,
    )

my_dialog_example3 = dialog_example3()
```

## Events

For dialogs, user defined callbacks should be chained with the close function in the `on_press` handler of the dialog's action buttons. The following example prints if the dialog's save or cancel button is clicked.

```python
from deephaven import ui

@ui.component
def print_example():
    is_open, set_open, set_closed = ui.use_flag()

    def print_save():

>         set_closed()

Reviewer 2: The Spectrum component passes a `close` callback to `on_press`. I'm assuming we opted not to implement this due to nuances of async funtions in dh ui, but I wanted to double check?

Author: Yes, I had a discussion with Matt and Bender about this on slack. The conclusion was to use controlled mode because of the difficulties involved in using the callback.

        print("Profile saved!")

    def print_cancel():
        set_closed()
        print("Provfile not saved!")

    return ui.dialog_trigger(
        ui.action_button("Set Profile", on_press=set_open),
        ui.dialog(
            ui.heading("Profile"),
            ui.content(

>                 ui.form(ui.text_field(label="Name"), ui.checkbox("Make private"))

Reviewer 2: For some reason, I can't type in the text field here which seems strange.

Author: Per your comment below, will look at https://github.com/deephaven/deephaven-plugins/issues/948

Reviewer 2: Yeah, this is unrelated to your PR. You can ignore.

            ),
            ui.button_group(
                ui.button("Cancel", variant="secondary", on_press=print_cancel),
                ui.button("Confirm", variant="accent", on_press=print_save),
            ),
        ),
        is_open=is_open,
    )

my_print_example = print_example()
```

### Dismissable dialogs

> Dismissable dialogs support an optional `on_dismiss` prop which is triggered whenever the dialog's close button is clicked. Similar to non-dismissable dialogs, you must chain the close function with whatever callback you provide as onDismiss. If this event callback is not needed, the dismissiable dialog will behave as normal without passing this callback through.

Reviewer: ```suggestion
Dismissable dialogs support an optional `on_dismiss` prop that is triggered whenever the dialog's close button is clicked. Like non-dismissable dialogs, you must chain the close function with whatever callback you provide as `onDismiss`. If this event callback is not needed, the dismissable dialog will behave normally without passing this callback through.
```

```python
from deephaven import ui

@ui.component
def dismissable_callback():
    is_open, set_open, set_closed = ui.use_flag()

    def print_dismiss():
        set_closed()
        print("Dialog dismissed.")

    return ui.dialog_trigger(
        ui.action_button("Info", on_press=set_open),
        ui.dialog(
            ui.heading("Version Info"),
            ui.content("Version 1.0.0, Copyright 2020"),
            on_dismiss=print_dismiss,
        ),
        is_open=is_open,
        is_dismissable=True,
    )

my_dismissable_callback = dismissable_callback()
```

## Visual options

### Dialog types

> Dialogs can be rendered as modals, popovers, or trays. See the `dialog_trigger` docs for more information.

Reviewer: link

Author: How do I link to one of the `dialog_trigger.md` doc which is in the same directory?

Author: Ok, I tried something. I think it works.
